### ISSUE TITLE
Why don't we use promises in our API?

### ISSUE THREAD

**Issue by jgeewax** [GoogleCloudPlatform/google-cloud-node#551]
#### Promises coming soon. Track progress here: #1142

---

Seems like we're passing callbacks in rather than getting promises back -- was this an explicit choice?

/cc @idosela

**Comment by stephenplusplus** [comment]
Why *should* we use promises? We talked about it before, but Node is callbacks, promises are an abstraction layer that not everybody enjoys using.

**Comment by jgeewax** [comment]
I'll leave it to @idosela to chime in on why he thinks they're great :)

I actually think they're nice, but don't think it's a make-or-break.

**Comment by ryanseys** [comment]
Yeah I think it'd be way too much work now (for us and for our users) to switch. That's a design decision definitely made up front. Also, all our dependencies and the large majority of npm packages in general use callbacks, so while I can understand the appeal for Promises for certain cases, they don't exactly sit well with what people come to expect from a Node library.

**Comment by jgeewax** [comment]
And I'm guessing that if someone wanted to, they could wrap what we have and add promises... I'm going to close this out as I think we've answered the question:

- Yes, we talked about it before, but it wasn't a consistent with the rest of the Node libraries out there.
- Yes, it's nice, but we probably won't add it.
- If someone really loves promises, they're welcome to fork and maintain their own wrapper?

Feel free to re-open @idosela if you think we're making a huge mistake, otherwise we'll continue with our trusty ol' callbacks :)

**Comment by jgeewax** [comment]
... and @silvolu just pointed me at https://www.npmjs.com/package/gcloud-promise (hosted at https://github.com/sramam/gcloud-promise)

Thanks @sramam ! :+1:

**Comment by idosela** [comment]
**TL;DR;** I like promises, but I understand it would take a lot of work to switch to promises, so feel free to leave the issue closed.

I find promises to be very convenient for handling async operations. The most common use cases in which promises provide significant advantage over callbacks are combination, chaining, and error handling.

Let's say I want to fire off 10 Datastore queries, and then do something when they are all done.
With callbacks I would have to write a bunch of code to track when each query is complete. With promises I can combine all of the promises returned by each individual query into a single promise which will be resolved when all of the queries are done. All it takes is a single function call.

Promise chaining allows you to break down sequential async operations which depend on each other's return values, without having to resort to nesting of callbacks. This leads to a more readable and testable code.

Promises also allow the user of your API to write less code, and have more flexibility. For example, with callbacks I constantly have to check whether an error happened, and can't have a single error handler for multiple related operations. 

**Callbacks**
```js
dataset.save({key: blogPostKey, data: blogPostData}, function(err, response) {
  // Handle error in step 1.
  if (err) {
    return;
  }

  dataset.save({key: blogPostKey, data: {isDraft: false}}, function(err, response) {
    // Handle error in step 1.
    if (err) {
      return;
    }

    // Do something on success.
  });  
});
```

**Promises**
```js
dataset.save({key: blogPostKey, data: blogPostData})
    .then(function(response) {
      return dataset.save({key: blogPostKey, data: {isDraft: false}});
    })
    .then(function(response) {
      // Do something on success.
    })
    .catch(function(err) {
      // Handle errors from step 1 or 2.
    });
```

Q (https://github.com/kriskowal/q) provides much more detailed comparison of callbacks and promises, and I know of several Node packages that use it. I'm no Node expert, but it seems to me Promises are a pattern that is not language or environment specific.

Sorry for going on and on :) Have a great weekend!

**Comment by jgeewax** [comment]
Out of curiosity -- is there a way we could easily support both ?

**Comment by jgeewax** [comment]
I'm reopening this issue mainly because ... I've been writing more stuff in Node and promises are becoming nicer and nicer.

Some libraries that use promises that seem really nice:

- http://docs.sequelizejs.com/en/latest/
- https://www.firebase.com/docs/web/libraries/angular/guide/user-auth.html#section-logging-users-in

I think our library would be... kind of amazing if we could offer promises as well as the callback style we have today.

If we were to invest in this, I'd love if it we could do the work to support promises, without documenting them, and once they are ubiquitous in the library, we post the documentation.

Thoughts?

**Comment by ryanseys** [comment]
What value does it add? What do you mean by "nicer and nicer"? Callbacks vs. Promises is mainly, I feel, just a personal preference. I don't like the idea of mixing both callbacks and promises because it can confuse people if they mix them. Do you know of any popular node libraries that support both callbacks and promises out of the box?

**Comment by jgeewax** [comment]
> What value does it add? What do you mean by "nicer and nicer"? Callbacks vs. Promises is mainly, I feel, just a personal preference.

I'd consider the nesting of callbacks to be an anti-pattern... Further, "we accept a callback" is rather restrictive -- and "we return a promise, do what you want" is much more open.

> I don't like the idea of mixing both callbacks and promises because it can confuse people if they mix them.

This could be true, but what's wrong with saying in the docs: "this method accepts a callback, and returns a promise" ? Does that really confuse you?

> Do you know of any popular node libraries that support both callbacks and promises out of the box?

@idosela : Do you know of any popular node libraries that support both? I don't know of any off the top of my head...

**Comment by sramam** [comment]
If all that is needed is a promise interface to gcloud-node,
a couple of lines and you're there!

Check out the excellent bluebird library and it's delightful
promisification capabilities:
https://github.com/petkaantonov/bluebird/blob/master/API.md#promisification

-shishir


On Mon, Jun 8, 2015 at 4:32 AM, JJ Geewax <notifications@github.com> wrote:

> What value does it add? What do you mean by "nicer and nicer"? Callbacks
> vs. Promises is mainly, I feel, just a personal preference.
>
> I'd consider the nesting of callbacks to be an anti-pattern... Further,
> "we accept a callback" is rather restrictive -- and "we return a promise,
> do what you want" is much more open.
>
> I don't like the idea of mixing both callbacks and promises because it can
> confuse people if they mix them.
>
> This could be true, but what's wrong with saying in the docs: "this method
> accepts a callback, and returns a promise" ? Does that really confuse you?
>
> Do you know of any popular node libraries that support both callbacks and
> promises out of the box?
>
> @idosela <https://github.com/idosela> : Do you know of any popular node
> libraries that support both? I don't know of any off the top of my head...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-node/issues/551#issuecomment-109957512>
> .
>



-- 
Imagine there were no hypothetical situations. ಠ_ಠ

**Comment by jgeewax** [comment]
Very cool -- is it crazy to either document this (how you promisify our stuff using this library) or actually provide some code that does this (taking a dependency on this library)?

My worry is that

- people will struggle to promisify stuff and get frustrated (so docs would help here)
- we will do something that breaks the promisification, and have scary bugs (adding code + tests would help here)

**Comment by sramam** [comment]
On Mon, Jun 8, 2015 at 6:12 AM, JJ Geewax <notifications@github.com> wrote:

> Very cool -- is it crazy to either document this (how you promisify our
> stuff using this library) or actually provide some code that does this
> (taking a dependency on this library)?
>
> My worry is that
>
>    - people will struggle to promisify stuff and get frustrated (so docs
>    would help here)
>
> Perhaps documentation might be an option, but strictly speaking that is
outside the scope of the project.
That said, if one is interested in promises and node, it's not really that
hard to stumble upon Bluebird.


>
>    - we will do something that breaks the promisification, and have scary
>    bugs (adding code + tests would help here)
>
> All promisify() requires is that the wrapped node function *"should
conform to node.js convention of accepting a callback as last argument and
calling that callback with error as the first argument and success value on
the second argument." *
Within those confines, the underlying library's features, bugs and warts do
not discriminate based on race, gender or interface! :)

 Basically all I'm saying is "It's exceptionally easy and safe to use the
Bluebird promisify interface". Please try it out.


>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-node/issues/551#issuecomment-109990167>
> .
>



-- 
Imagine there were no hypothetical situations. ಠ_ಠ

**Comment by jgeewax** [comment]
> Perhaps documentation might be an option, but strictly speaking that is outside the scope of the project.

Sorry -- I meant that **we** would document how you would use bluebird. Not bluebird itself! :)

> Basically all I'm saying is "It's exceptionally easy and safe to use the Bluebird promisify interface". Please try it out.

Totally - what I was saying here is that if we do add support for this, we should have tests for it so that our users aren't the ones figuring out that things are broken. We should know when we make changes if it breaks our bluebird-enabled promise usage.

**Comment by kudos** [comment]
Promises allow this kind of syntax with modern javascript:

```js
var file = yield bucket.file('path/to/file').download();
```

This syntax is already possible in node/iojs without the help of any transpilers such as babel or traceur. It's also possible to support both styles simultaneously, by switching to a promise API when no callback is passed. The Stripe node library is an example of this.

**Comment by stephenplusplus** [comment]
> This syntax is already possible in node/iojs without the help of any transpilers such as babel or traceur. 

Node only supports generators behind the `--harmony` flag.

In some places in our API, we use the missing callback to enable stream mode:

```js
bucket.getFiles(function(err, files) {});
bucket.getFiles().on('data', function(file) {});
```

I would rather see the promisifying live in another module, so everything is clear for the maintainers and the users. An example would be [request](https://github.com/request/request) and [request-promise](https://github.com/request/request-promise.

**Comment by jgeewax** [comment]
I have to say, the Stripe way of doing this is really nice... And it doesn't look like (from my limited perspective) the stream vs promise thing is all that problematic... right?

For example, 

```javascript
// .on('data') returns a new promise.
bucket.getFiles().on('data', function(file) {
  // this is called on each file in the list
  done(); // or complete() or whatever it is...
}).then(function() {
  // this is called when the stream is complete
});

bucket.getFiles().then(function(files) {
  // handles files in chunks/pages/whatever
});

bucket.getFiles(function(files) {
  // same as above.
});
```

**Comment by stephenplusplus** [comment]
The stream api uses .on in a chaining manner which returns the stream to allow further piping. We would be hacking native APIs to do something magical to support 3 use cases. It's too much for our library. We can add a promise module pretty easily in the manner of request-promise, and users can just use that to get their promise on.

**Comment by jgeewax** [comment]
Are we worried at all about gcloud-promise lagging gcloud ?

**Comment by stephenplusplus** [comment]
When we add new methods or APIs here, it would likely be a deliberate action to add support there. Can't say for sure though, I'd have to think a little more on how we would do this. I think it would be safe to make waterfall releases, where we can add in anything necessary to the promise library immediately after a release here. And as long as we follow semver, no surprise breakages would arise.

@callmehiphop any thoughts on the best way to add support for promises?

**Comment by jgeewax** [comment]
Another option to toss out is just to have a separate test suite for
https://github.com/petkaantonov/bluebird, so that we know if something
isn't compatible or if we broke it....?

**Comment by callmehiphop** [comment]
@stephenplusplus I think supporting callbacks, promises and streams in a single API is probably possible, but is very likely to end up being kind of hacky. A separate project leveraging Bluebird is probably the cleanest way to go.

**Comment by jmdobry** [comment]
For what it's worth, the very first thing I did after pulling gcloud-node into my project was to figure out how to promisify it.

**Comment by jgeewax** [comment]
Interesting. Did you use bluebird? Or something else?

**Comment by jmdobry** [comment]
Bluebird. _Is_ there anything else? ;)

**Comment by kudos** [comment]
There really is no need for Bluebird. I'd much rather use native promises with a simple polyfill for legacy node.

**Comment by callmehiphop** [comment]
Our main problem with supporting promises is that a number of our methods return a stream if a callback is not provided.

@stephenplusplus feel free to tell me this is the worst idea ever, but what if we hid streams behind a different method/option?

```javascript
bucket.getFiles({ stream: true }).on('data', function(file) {});
```

My rationale is that promises could be leveraged almost everywhere in our api, where as we really only use streams in paginated lists.

**Comment by stephenplusplus** [comment]
Sure.

**Comment by jmdobry** [comment]
I _was_ going to submit a PR where Bluebird would be used internally with the `.nodeify(cb)` option so the API would support Promises and callbacks, but then I saw the streams and I was like "oh...guess I'll just comment on the thread first".

> different method/option?

I would argue for a different method rather than a different option, but that might involve breaking changes. What's the versioning/compatibility philosophy here?

**Comment by callmehiphop** [comment]
I believe the rationale was that we haven't hit 1.0 yet, so we're ok with breaking changes.

**Comment by chrishiestand** [comment]
+1 - promises and async await are well on their way to first class js features so I think libraries should use promises instead of callbacks whenever possible - which is afaik all async calls that aren't stream-like. So my vote is to drop callback support to keep your library simpler.

I've written a couple of helper libraries that bluebird promisify datastore and pubsub functions and I can say that while promisification isn't hard, it's not nearly as easy as not having to roll your own. It might be worth noting here that [bluebird 3 made some small but breaking api changes](http://bluebirdjs.com/docs/new-in-bluebird-3.html) that people writing wrappers should know about.

**Comment by stephenplusplus** [comment]
Thanks for doing that :)

Promises seem to be the way to. I think we should switch to promises and ".stream"-extended methods where they apply. No callbacks.

**Comment by callmehiphop** [comment]
I think if we move streams to their own method (e.g. `getBuckets.stream()`), then supporting both callbacks and promises would be pretty simple. I'm definitely in favor of promises, but I don't see why we need to drop callbacks altogether. That's probably going to make upgrading for a lot of people more difficult than it needs to be, especially if we're not going to bump the major version.

I'm in favor of putting streams behind a `.stream` method and supporting callbacks and promises with the idea that we'll eventually phase out callbacks in the future.

**Comment by stephenplusplus** [comment]
Pre-1.0 minor bumps are semantically equivalent to post-1.0 major bumps. I
just don't think supporting and documenting multiple ways to do the same
thing is necessary, only confusing. If we intend to phase it out, I don't
see why we don't just cut it outright.

**Comment by kudos** [comment]
@chrishiestand it doesn't make sense to drop callbacks for promises, since there's no conflict there. The only conflict is with streams and promises.

@callmehiphop's solution makes most sense to me, I disagree with @stephenplusplus that callbacks should be removed right now, though I could see that making sense at a later date.

**Comment by stephenplusplus** [comment]
Give us the promises, @callmehiphop :)

**Comment by callmehiphop** [comment]
My pleasure!

**Comment by sindresorhus** [comment]
I would go all in with promises. It's time. No point in making your API more complicated than necessary by supporting both. I personally use the tiny [pinkie-promise](https://github.com/floatdrop/pinkie-promise) Promise polyfill and [pify](https://github.com/sindresorhus/pify) to promisify callback-API. Really no need to depend on Bluebird in reusable modules.

Callback users can still easily use your API:

```js
promise().then(result => cb(null, result)).catch(cb);
```

**Comment by jmdobry** [comment]
> I would go all in with promises. It's time.

I agree.

ES6 Promises are becoming more widely available, as seen [here](https://kangax.github.io/compat-table/es6/#test-Promise). gcloud-node could just assume `global.Promise` is available, and it's easy enough for a user to polyfill a legacy environment if necessary.

**Comment by waprin** [comment]
+1 to jmdbory and sindresorhus

**Comment by Muqito** [comment]
I would love this as well

    gcs.createBucket(bucketName).then(function(bucket) {
        return bucket.upload(folder);
    }).then(function() {
        // folder is uploaded
    });

+1 for promises

**Comment by callmehiphop** [comment]
It sounds like our plan is to go all in with promises, it's a pretty big change so we're going to wait until at least early January to switch.

**Comment by jasonswearingen** [comment]
A bit surprising that promises are not supported.  I would suggest supporting promises and callbacks, but it might get in the way with the various Stream-returning overloads

**Comment by stephenplusplus** [comment]
We're getting there, @jasonswearingen. Promise.

**Comment by jgeewax** [comment]
I see what you did there....... :)

**Comment by DeeSnow97** [comment]
I didn't exactly have time to read this debate about whether to convert the entire API to promises or not, but a few days ago while working with this API I encountered the same problem. I'm actually writing the entire app with promises, including gcloud calls. It's not hard to add some support manually, let me share my quick and dirty hack with you:

```javascript
dataset.promise = function( method ) {
	var args = Array.prototype.slice.call( arguments, 1 );
	return new Promise( function( resolve, reject ) {
		dataset[method].apply( dataset, args.concat([ function( err ) {
			if( err ) {
				reject( err );
			} else {
				resolve( Array.prototype.slice.call( arguments, 1 ) );
			}
		}]));
	});
};
```
Then you can just call it like `dataset.promise('get', ['User', userid])` which returns a promise. If there is no error, it's resolved to an array of the original arguments except the first one, and if the first argument indicates an error, it fails and you can handle the error with `catch()`. And it changes no existing functionality, so old code using callbacks won't be affected. 

Also, this one is specific to the dataset interface, but it can be easily extended to any interface assuming the last argument is the callback and the first argument of the callback is the error variable:
```javascript
[
	dataset
].map( function( object ) {
	object.promise = function( method ) {
		var args = Array.prototype.slice.call( arguments, 1 );
		return new Promise( function( resolve, reject ) {
			object[method].apply( object, args.concat([ function( err ) {
				if( err ) {
					reject( err );
				} else {
					resolve( Array.prototype.slice.call( arguments, 1 ) );
				}
			}]));
		});
	};
});
```
Just add your interfaces to the array and it implements the promise way. Thought it could be useful.

Example usage right from my code:
```javascript
function register( data ) {
	var key = dataset.key(['User']) ;
	var data = {
		fbid: data.fbid,
		name: data.name,
		avatar: data.avatar
		//... and some more stuff
	};
	return dataset.promise( 'save', {
		key: key,
		data: data
	}).then( function( response ) {
		data.id = key.id ;
		delete data.fbid ;
		return data ;
	});
}
```
(This function is also used like `register().then().then().then().catch()`)

**Comment by stephenplusplus** [comment]
Thanks for providing that! To catch you up, we are implementing promises, we just haven't gotten to it yet.

We're 75% settled on using arrays to resolve promises with multiple parameters, but feel free to weigh in on how you'd like to see this situation handled:

```js
// Callbacks
firewall.delete(function(err, operation, apiResponse) {});

// Promises (style 1)
firewall.delete()
  .then(function(response) {
    var operation = response[0];
    var apiResponse = response[1];
  }, function(err) {});

// Promises (style 2)
firewall.delete()
  .then(function(response) {
    var operation = response.operation;
    var apiResponse = response.apiResponse;
  }, function(err) {});
```

Anyone using ES6 gets destructuring either way:

```js
firewall.delete()
  .then(function([operation, apiResponse]) {
    // operation & apiResponse are ready to use
  });
```

But by using arrays, we are compatible with Bluebird users, who have [`spread`](http://bluebirdjs.com/docs/api/spread.html) available to them.

**Comment by stephenplusplus** [comment]
Alas, a PR! #1142 -- anyone interested, please subscribe there to track our progress. Thanks for all of your input so far!

**Comment by jasonswearingen** [comment]
@stephenplusplus regarding "no callbacks, only promises".   This is fine with me as I am also "all in" on promises, 

but FYI though some opinions here are from capable and influential members of the Node community, the are rather opinionated about supported features.  Some of the more "retro" nodejs users who haven't switched to Promises for whatever reason are going to have greater challenges using ```gcloud-node``` because of this,   and it seems like there is a very little design/implementation impact in having optional callbacks .

**Comment by stephenplusplus** [comment]
I definitely agree, but I don't think our library should support multiple ways to do the same thing. I still think this: https://github.com/GoogleCloudPlatform/gcloud-node/issues/551#issuecomment-153177623

**Comment by jasonswearingen** [comment]
ok, as a workaround, you can also mention in the docs that people who don't use promises can use the v0.28 version of the api's.

**Comment by jgeewax** [comment]
Eh. Yikes.

What if we only support one way (cb) but then make bluebird work perfectly
well...? Is that a terrible direction? I can't recall what we said about
that.

**Comment by kudos** [comment]
@jgeewax I don't want to use Bluebird, my runtimes (node, browser) all have native promises.

**Comment by jasonswearingen** [comment]
@kudos fyi, most browsers don't have native promises (yet) as that's es6.  Also bluebird is (just) a A+ compatible promise library.  being compatible with bluebird means being compatible with promises in general.

**Comment by jgeewax** [comment]
:) Thanks @kudos . Just want to make sure we're doing "The Right Thing (TM)".

@jasonswearingen : I really want to get a better understanding of those people who are influential to the community, but don't want "only promises"... Any chance anyone tuned in can help me wrap my head around this one?

I think that the general summary so far is...

- we could make everyone happy by doing both (cb + promises)
- doing both will likely be a pretty big headache for us (right, @stephenplusplus ?)
- doing just callbacks means we could have promises via Bluebird
- doing just promises means that people need to type `method().then(cb)` instead of `method(cb)`

**Comment by kudos** [comment]
@jasonswearingen pretty sure Bluebird is being brought up in the context of using its `promisifyAll` method, and then not implementing promises in this library at all.

**Comment by jgeewax** [comment]
@kudos: Yes, that was what I meant. Do just callbacks, and then let people who want promises use Bluebird. No idea if that's the right way to go, but it is something to discuss. Which @stephenplusplus I think may have already settled... :-x

**Comment by kudos** [comment]
@jgeewax the original issue was a clash with the streams API. Here's a quick example of how it could conceivably work from a single API:

```js
method(arg, cb);
method(arg).then(cb);
var something = yield method(arg);
method(arg, {stream: true}).on('data', cb);
```

Error handling is left as an exercise for the reader :P

**Comment by stephenplusplus** [comment]
I'm happy with any solution @callmehiphop thinks will work. I only want what the community wants :)

**Comment by jasonswearingen** [comment]
@jgeewax the problem with the ideals of "those influencial people" is they tend to be eager adopters of the latest+greatest tech, which really bites the lagers in the butt.  For Promises, nodejs doesn't require (or support) them in the core api's, so it's requirement is a barrier-to-entry for those who don't use/understand promises.   

Another example (not related to gcloud-node obviously) is a lot of microlibraries these days are moving their codebases to ES6, which sounds great but is exceedingly hostile to use in browsers.   (requiring the consumer to know about transpiling).

I am all in favor of Promises myself, but answering support questions for my products I am keenly aware of....  "less sophisticated" professional developers who struggle to even make a proper POST request.   So I'm just trying to advocate a little flexibility for their benefit.

**Comment by jgeewax** [comment]
Thanks @jasonswearingen -- really appreciate the context. 

To be fair, I have to double check the right flags in cURL for POST requests... :disappointed: But I see the point.

**Comment by callmehiphop** [comment]
Personally I'm not very into the idea of optimizing our library for Bluebird use. I think with native async functions on the horizon, it would well worth our time to offer Promise support natively from within our library. As @kudos pointed out, the blocking issue here is that it's not currently feasible to differeniate between streams and promises with our current structure. As suggested throughout this thread, promises would likely require us to moving all streaming APIs behind a `.stream` method.

To @jasonswearingen's point, going all in with Promises will potentially make it difficult for some of our customers to upgrade, which is the only thing that has made me hesitant to go *all in* with Promises.

That being said, I'd like to piggy back to my previous proposal of moving streaming APIs behind a `.stream()` method and supporting both Promises *and* callbacks with the intention of possibly phasing out callbacks in the future.

**Comment by jasonswearingen** [comment]
regarding streams, there is also the high quality (and reasonably popular) RxJs observables library, which is basically "Promise+Streams".  However it's enterprise C# roots show, so it might not be something you "go all in on", but worth keeping on your radar.

https://github.com/Reactive-Extensions/RxJS

**Comment by chrishiestand** [comment]
After some discussion with a pro-callback friend I now think that gcloud-node should provide both callbacks and promises. There are two problems with switching to only promises, as I see it:

1. Promises, without async/await (which *might* be in ES2017), can still be a bit awkward to use. For example, there's no clean way to "return early" from a then() chain.
2. Developers currently use all sorts of tools to handle async workflows, like rxjs (as mentioned above), or js-csp. These were developed prior to Promises becoming standard and many developers still prefer them.

So until the community has by and large adopted Promises + async/await (which could conceivably not happen if some other paradigm takes root), I think we should offer both.

This sort of dual-support is relatively common. Here are some libraries I've used which do this:
https://github.com/elastic/elasticsearch-js
https://github.com/luin/ioredis

How easy it is to implement here, I don't know.

**Comment by cryptoquick** [comment]
As a regular Joe Schmoe application developer who likes to use Babel tooling in my projects, I'd appreciate a promise-based API because async and await have proven to be really nice and convenient, and I've been using them extensively through throughout my projects, gluing together callback-based APIs where necessary with the ES6 Promise object and its resolve/reject handlers.

What I've seen many APIs do is to simply provide a promise if no callback argument is passed. Conceivably, this would not be a breaking change.

**Comment by jasonswearingen** [comment]
I'm not sure if this is the appropriate place to suggest a design, but for **datastore transaction promises**:

I wrote a promise wrapper over the Datastore portion, and think I have a pretty reasonable transaction signature:

```
dataset.runInTransaction((/**can pass transaction to read/write methods as normal */transaction)=>{
//return resolved promise to commit transaction
//return rejected promise or throw error to rollback
})
//returns a promise that resolves when the transaction is finished.
```

**Comment by c0b** [comment]
> I don't think this is our top priority at the moment, so it's hard to nail down an accurate estimate, but I would guess sometime within the next several months.

but why is it so slow, look at AWS SDK has a better support of Promise API /cc @callmehiphop

https://blogs.aws.amazon.com/javascript/post/Tx3BZ2DC4XARUGG/Support-for-Promises-in-the-SDK
http://docs.aws.amazon.com/AWSJavaScriptSDK/guide/node-making-requests.html#Support_for_Promises

**Comment by rommguy** [comment]
Until promises are supported, if someone wants to use promises, you can use bluebird method promisifyAll, which converts a node callback based API to a promise based one. Lots of examples with fs module, but I use it with the datastore too.

**Comment by callmehiphop** [comment]
Great news everyone, Promises are officially out as of [v0.42.2](https://github.com/GoogleCloudPlatform/google-cloud-node/releases/tag/v0.42.2)! Thanks to everyone who contributed to this, I know it was a highly sought after feature and I'm happy that it's finally in!
